File: app/src/main/java/fi/kaupunkifillarit/inject/KaupunkifillaritComponent.java
package fi.kaupunkifillarit.inject;

import dagger.Component;
import fi.kaupunkifillarit.KaupunkifillaritApplication;

@ApplicationScope
@Component(modules = {SystemServicesModule.class, ServicesModule.class})
public interface KaupunkifillaritComponent extends KaupunkifillaritGraph {
    /**
     * An initializer that creates the graph from an application.
     */
    final class Initializer {
        public static KaupunkifillaritComponent init(KaupunkifillaritApplication app) {
            return DaggerKaupunkifillaritComponent.builder()
                    .systemServicesModule(new SystemServicesModule(app))
                    .servicesModule(new ServicesModule(app))
                    .build();
        }

        private Initializer() {
        } // No instances.
    }
}

File: app/src/main/kotlin/fi/kaupunkifillarit/inject/KaupunkifillaritComponent.kt
package fi.kaupunkifillarit.inject

import dagger.Component
import fi.kaupunkifillarit.KaupunkifillaritApplication
import fi.kaupunkifillarit.MapActivity
import javax.inject.Singleton

@Singleton
@Component(modules = arrayOf(SystemServicesModule::class, ServicesModule::class))
interface KaupunkifillaritComponent {
    fun inject(application: KaupunkifillaritApplication)

    fun inject(activity: MapActivity)

    /**
     * An initializer that creates the graph from an application.
     */
    object Initializer {
        fun init(application: KaupunkifillaritApplication): KaupunkifillaritComponent {
            return DaggerKaupunkifillaritComponent.builder()
                    .systemServicesModule(SystemServicesModule(application))
                    .servicesModule(ServicesModule(application))
                    .build()
        }
    }
}

